### DISCUSSION TITLE
Static site generation (SSG) from niceGUI

### DISCUSSION THREAD

**Discussion by linuxiamo** [zauberzeug/nicegui#637] [Ideas / Feature Requests]
Hello,
As by title, there will ever be the possibility to compile a static site from niceGUI?
(If already possible, could you please address me to a guide on how to do it?)
Thanks for the attention.

**Comment by rodja** [comment]
As said in a discussion [over at Reddit](https://www.reddit.com/r/nicegui/comments/11wl7op/comment/jd6t8fk/?utm_source=share&utm_medium=web2x&context=3), NiceGUI follows a "backend first" strategy, which means that it generates the frontend on the fly through the backend and can not provide a static version of the code. But as mentioned in #451 it may be possible to export server side rendered static files (SSR) that could still communicate with a server backend. But adding this feature may be quite complex.

For what purpose would you use static site generation?

**Reply by linuxiamo** [reply]
Hi, thanks for the answer. Apologize me for the delayed reply. I asked if it was possible, because I needed to create a minimal site to share some information and some simple blog articles. I know it is a request outside the main scope of niceGUI, but that was my need and preferably, I wouldn't bother building it with other dedicated tech stacks at the moment.

**Comment by falkoschindler** [comment]
Possibly related discussion: #313

**Comment by legout** [comment]
My two cents... :-)

As already discussed on [reddit](https://www.reddit.com/r/nicegui/comments/11wl7op/comment/jdm6pee/?context=3) exporting nicegui apps to (prerendered) static sites would be a great feature, which might enhance SEO and performance, according to

https://jamstack.org/ or
https://www.freecodecamp.org/news/how-to-make-static-site-dynamic/

nuxt is a web framework for vue.js apps, which allows for static site export. https://nuxt.com/docs/getting-started/deployment#static-hosting

Is it possible to make nicegui compatible to nuxt?

**Comment by rodja** [comment]
I just realized the difference between server-side rendering (SSR) and static-site generation (SSG):

- With SSR, the HTML is generated on-the-fly on the server for each incoming request.
- With SSG, the HTML is generated once during the build process, and static files are served for each request.

To decide which tech we pursue we first need to nail down the requirements. I think the main goal is to get

1. better SEO
2. faster loading times
3. improved "social sharing" previews 

I feel that things like

4. serving via CDN
5. hosting on GitHub Pages, Netlify, Vercel, ...

are not so important for NiceGUI because of our backend-first strategy. The main benefit of NiceGUI is controlling the behavior of the frontend via backend. When putting pre-generated html on GitHub Pages or similar there is no backend anymore which makes NiceGUI ineffective.

Do all agree? Then we should change the title of this feature request to SSR and focus on how this could be added. To send the initial content ready-made to the client for better SEO and performance while still keeping real-time updates via websocket, we can use [Server-Side Rendering (SSR) combined with client-side hydration](https://quasar.dev/quasar-cli-vite/developing-ssr/configuring-ssr#introduction). This approach will enable us to deliver pre-rendered HTML to the client, which can then "hydrate" into a fully functional Vue3 app that still allows NiceGUI perform real-time updates.

**Reply by falkoschindler** [reply]
I'd vote against renaming this discussion. The original proposal was to provide static side generation, and I just pointed @legout to this thread who proposed the very same at https://github.com/zauberzeug/nicegui/discussions/645.
Let's create a new discussion for SSR and leave this one here as a landing page for SSG - even if we come to the conclusion to not pursuit it with priority.

**Comment by dclause** [comment]
My two cents: I can't figure out at the moment a way around SSR. Indeed, niceGUI builds a page with various technics (JSX files, SFC files and render() components) that will be feed to vueJS to handle the rendering client side. 
What Nuxt does for instance is to actually run vueJS in a node environment. That is JS code (vueJS) running on the server, just as it would be on the client. 
NiceGUI don't have this opportunity because it runs on a fully different tech stack (namely Python) so it can't run the vueJS code in this environment. 
I can't make my mind around a way to do that without using a Node stack in parallel which would (on my humble opinion) not be a good idea. 

However to improve performance we can: load the bare-minimum code from the server (see my proposal on PR #658), remove as much external dependencies as possible (remove aggrid and plotly to depend on ui.table and python mathplot?, maybe tailwind) and discharge  the server for unecessary load (use CDN for all front resources) for instance. 
We could imagine adding page preload, etc... 

It all depends on niceGUI use case I guess and what target it has. If its purpose is to run on lite (raspberry for instance) embedded systems, then the browser on your computer (and your client network) is actually probably more capable .

**Reply by rodja** [reply]
@dclause I just created a separate discussion for SSR to not mix it with SSG: #663. Would you like to repost your thoughts there? Then we have everything in its proper place (as @falkoschindler suggested in https://github.com/zauberzeug/nicegui/discussions/637#discussioncomment-5448349).

**Comment by cj** [comment]
Another solution could be creating a python script to generate the static files. You give it an array of all the pages you wanted to be compiled to a static site; you then have it grab all the rendered HTML from it, fetch and save all the external JS, CSS, and image files, replace the links on the page to statically saved ones and then upload all those files to a static site host like Netlify or https://render.com/docs/static-sites.

I plan on creating something along those lines. I'll link the code here once I'm done :)

**Reply by rodja** [reply]
Interesting. You plan on fetching the rendered HTML from the browser? Maybe you want to look at our pytests. We use headless selenium as described in [tests/README.md](https://github.com/zauberzeug/nicegui/blob/main/tests/README.md).

**Reply by cj** [reply]
@rodja exactly, there used to be a service for this https://sitesauce.app. There are also some other open source clis, when I find them again I'll paste them in here.

**Comment by rodja** [comment]
We had another inquiry for exporting to static html on Reddit: https://www.reddit.com/r/nicegui/comments/13d14a5/nicegui_static_site_optionexport/

**Comment by rodja** [comment]
The latest Reddit post kept me thinking... maybe someone wants try using https://brython.info/ (or another Python-in-the-Browser solution) to run a variant of NiceGUI on the client machine. Similar to the request of being able to run on Django (see https://github.com/zauberzeug/nicegui/discussions/528), NiceGUI may one day support different kind of backends. And one of them could be "browser-based"....

**Comment by sjvdm** [comment]
I have been using nicegui for quite a bit of backend stuff - then had to build a simple frontend. I am not a JS expert, so doing this in nicegui was super quick and easy - but hosting this on a service that supports python (vs something like firebase etc) seems like overkill.

If I find an easy-ish solution, I will post it here.

On the haters of following this approach - there are billions of languages out there, and building something on a language you know (aka Python) vs something you don't (JS) shortens the dev cycle considerably!

**Reply by samuller** [reply]
If you don't need any interactivity beyond static links and just want to use NiceGUI to generate static pages, then you could maybe host a server, connect to it and save the HTML that was generated for that initial connection. You'll then also need the proper folder structure with dependencies, etc.

I'm not sure if this is a good idea or even if attempting to use NiceGUI for this goal makes proper sense, but below I have created a script that does the above. It seems to work for NiceGUI 3.4.0 and can be used as follows:
```bash
# Create UI and generate HTML from initial connection which is then saved in new folder "generated_html"
uv run python generate.py
# Go into folder and host its static content
cd generated_html
python3 -m http.server 8000
```

<details>
    <summary>Script to generate static HTML</summary>

```python
import asyncio
import contextlib
import os
import logging
from pathlib import Path
import shutil
from typing import NamedTuple, cast
import uuid

from fastapi import Request
import httpx
from nicegui import Client, app, dependencies as ng_deps, ui, __version__ as ng_version
from nicegui.testing import User


logging.basicConfig(format="[%(levelname)s] %(message)s")
logger = logging.getLogger(__name__)
logger.setLevel(logging.INFO)


class StaticFiles(NamedTuple):
    url_path: str
    local_directory: str


@contextlib.contextmanager
def set_env_var(key, value):
    """Context manager for temporarily setting environment variable."""
    original_value = os.environ.get(key)
    os.environ[key] = value
    try:
        yield
    finally:
        if original_value is None:
            del os.environ[key]
        else:
            os.environ[key] = original_value


async def ui_generate(
    routes: list[str] | None = None,
    output_dir="generated_html",
    static_files: list[StaticFiles] | None = None,
    port: int = 8080,
    disable_websocket: bool = True,
    **kwargs,
):
    """
    Generate static HTML by running a NiceGUI server and saving the HTML generated upon initial request.

    NOTE: The generate HTML can't be opened directly in the browser, you have to host it with a web server, e.g.
              python3 -m http.server 8001
          Except it also needs to handle the MIME type of `.vue` files correctly (e.g. as "application/javascript").

    Parameters
    ----------
    routes
        List of route paths for which to generate HTML pages.
    output_dir
        Where to save generated static HTML.
    static_files
        List of tuples for static files that your app added using `app.add_static_files()`.
    port
        An available port on which to run internal server used for generation.
    disable_websocket
        Disable initial websocket connection attempt.
    kwargs
        Arguments passed to ui.run().
    """
    routes = routes or ["/"]
    output_dir = Path(output_dir)
    static_files = static_files or []

    # We use this envvar so we can register ui.pages and routes without fully running the server
    with set_env_var("NICEGUI_USER_SIMULATION", "1"):
        ui.run(
            storage_secret=str(uuid.uuid4()),
            port=port,
            **kwargs,
            # Variables that can't be configured by kwargs
            show=False,
            on_air=None,
            native=False,
            fullscreen=False,
            frameless=False,
            reload=False,
        )

        async with app.router.lifespan_context(app):
            async with httpx.AsyncClient(
                transport=httpx.ASGITransport(app), base_url=f"http://localhost:{port}"
            ) as http_client:
                # TODO: rather scan through all non-dynamic routes?
                # for app_route in app.routes:
                #     logger.info(app_route.path, app_route)
                # logger.info(Client.page_routes)

                for route in routes:
                    client: Client = await User(http_client).open(route)

                    request = Request(
                        # from starlette.types import Scope
                        {
                            "type": "http",
                            "headers": [],
                            "path": route,
                            # Set by SessionMiddleware
                            "session": {"id": "12314"},
                        }
                    )
                    # Generate HTML
                    response = client.build_response(request)

                    # Save generated HTML
                    logger.info("Generating...")
                    output_html_dir = output_dir / route.removeprefix("/")
                    output_html_dir.mkdir(parents=True, exist_ok=True)
                    with open(output_html_dir / "index.html", "wb") as file:  # noqa: ASYNC230
                        response_body = response.body
                        ########
                        # Temporary hack - remove all lines referencing vue components due to import and Mimetype issues
                        # (might be fixed in 3.5.0 after PRs 5567 and 5568?)
                        ########
                        remove_components = ["joystick", "plotly"]
                        filtered_lines = [
                            line
                            for line in response_body.decode("utf-8").splitlines()
                            # We exclude the "importmap" line by looking if it also contains xterm
                            if all(cmp not in line for cmp in remove_components) or "xterm" in line
                        ]
                        response_body = "\n".join(filtered_lines) + "\n"
                        ########
                        file.write(response_body.encode("utf-8"))

    # Alternative attempt to generate HTML without starting server
    # from nicegui.storage import request_contextvar
    # request_contextvar.set(request)
    # with Client(ui.page(route), request=request) as client:
    #     generate_page(request)
    #     # ui.label("TEST")

    # Copy NiceGUI static files
    ng_dir = Path(ui.__file__).parent
    output_ng_dir = output_dir / "_nicegui" / ng_version
    output_ng_dir.mkdir(parents=True, exist_ok=True)
    logger.info(f"Copying static files... {ng_dir / 'static'}")
    shutil.copytree(src=ng_dir / "static", dst=output_ng_dir / "static", dirs_exist_ok=True)
    if disable_websocket:
        # Delete socket.io to prevent websocket connection attempts
        os.remove(str(output_ng_dir / "static" / "socket.io.min.js"))
    # Copy per-component static files
    for url_path, resource_type in cast(
        list[tuple[str, dict]],
        [
            ("components", ng_deps.js_components),
            ("components", ng_deps.vue_components),
            ("libraries", ng_deps.libraries),
            ("resources", ng_deps.resources),
            ("esm_modules", ng_deps.esm_modules),
        ],
    ):
        resource_type_items = list(resource_type.items())
        logger.info(f"Copying {len(resource_type_items)} {url_path}...")
        for path, component in resource_type_items:
            dest = output_ng_dir / url_path / Path(path)
            dest.parent.mkdir(parents=True, exist_ok=True)
            logger.info(f"Copying {component.path} {url_path}...")
            if Path(component.path).is_dir():
                shutil.copytree(src=component.path, dst=dest, dirs_exist_ok=True)
            else:
                shutil.copy2(src=component.path, dst=dest)
    # Copy user-defined static directories
    for url_path, local_dir in static_files:
        shutil.copytree(src=local_dir, dst=output_dir / url_path.removeprefix("/"), dirs_exist_ok=True)

    # TODO: automatically detect routes defined with app.add_static_files()
    # for app_route in app.routes:
    #     if app_route.name == "static_file":
    #         logger.info(app_route.path, app_route)


def example_ui():
    # We create the UI directly here as an example, but for your own code you should probably import the Python
    # modules containing @ui.pages here so that the routes are setup, e.g.
    #     from my_custom_ui import routes
    @ui.page("/")
    def home_page():
        ui.label("This is home")
        ui.checkbox("You can't check me:")
        ui.link("About", target="/about")

    @ui.page("/about")
    def about_page():
        md_content = """
        ### About page
        * <i class="q-icon material-icons">person</i> Person
        * <i class="q-icon material-icons">phone</i> Phone

        ---

        Loading bar:
        """
        with ui.column().classes("w-full md:w-3/4 lg:w-[640px] items-stretch mx-auto justify-center"):
            with ui.card().classes("pt-0"):
                ui.markdown(md_content).classes("w-full [&_ul]:list-none [&_ul]:pl-5")
                ui.linear_progress(show_value=False).props("indeterminate")


if __name__ == "__main__":
    example_ui()

    static_files: list[StaticFiles] = []
    # If your app used static files, e.g. `app.add_static_files("/static", static_path)` then you would need to add
    # them here, e.g.
    #     # Define path to our local static directory
    #     static_path = Path(routes.__file__).parent / "resources" / "static"
    #     static_files.append(StaticFiles("/static", str(static_path)))

    asyncio.run(ui_generate(["/", "/about"], output_dir="generated_html", static_files=static_files))
```
</detail>

**Reply by sjvdm** [reply]
@samuller , this is awesome - you are a legend. I will let you know when I have tested it after the new year's rush!
